astropy/utils/data.py

```diff
@@ -1311,79 +1426,459 @@ def _get_download_cache_locs():
+            if not os.path.exists(h):
+                nonexistent_targets[u] = h
+            leftover_files.discard(h)
+            d, hexdigest = os.path.split(h)
+            if dldir != d:
+                bad_locations[u] = h
+            if check_hashes:
+                hexdigest_file = compute_hash(h)
+                if hexdigest_file != hexdigest:
+                    bad_hashes[u] = h
+        for h in leftover_files:
+            h_base = os.path.basename(h)
+            if len(h_base) >= 32 and re.match(r"[0-9a-f]+", h_base):
+                abandoned_files.add(h)
+        leftover_files -= abandoned_files
+
+        msgs = []
+        if nonexistent_targets:
+            msgs.append(
+                f"URL(s) point(s) to nonexistent file(s): {nonexistent_targets}")
+        if bad_locations:
+            msgs.append(
+                f"URL(s) point(s) to file(s) outside {dldir}: {bad_locations}")
+        if bad_hashes:
+            msgs.append(
+                f"Filename(s) does not match hash(es) of contents: {bad_hashes}")
+        if abandoned_files:
+            msgs.append(
+                f"Apparently abandoned files: {abandoned_files}")
+        if msgs:
+            raise CacheDamaged(
+                "\n".join(msgs),
+                bad_urls=(list(nonexistent_targets.keys())
+                          + list(bad_locations.keys())),
+                bad_files=(list(bad_hashes.values())
+                           + list(abandoned_files)))
         else:
-            return
-    msg = ("Unable to acquire lock for cache directory ({0} exists). "
-           "You may need to delete the lock if the python interpreter wasn't "
-           "shut down properly.")
-    raise RuntimeError(msg.format(lockdir))
+            return leftover_files
 
 
-def _release_download_cache_lock():
-    lockdir = os.path.join(_get_download_cache_locs()[0], 'lock')
+def import_file_to_cache(url_key, filename,
+                         hexdigest=None,
+                         remove_original=False):
+    """Import the on-disk file specified by filename to the cache.
 
-    if os.path.isdir(lockdir):
-        # if the pid file is present, be sure to remove it
-        pidfn = os.path.join(lockdir, 'pid')
-        if os.path.exists(pidfn):
-            os.remove(pidfn)
-        os.rmdir(lockdir)
-    else:
-        msg = 'Error releasing lock. "{0}" either does not exist or is not ' +\
-              'a directory.'
-        raise RuntimeError(msg.format(lockdir))
+    The provided ``url_key`` will be the name used in the cache. The file
+    should contain the contents of this URL, at least notionally (the URL may
+    be temporarily or permanently unavailable). It is using ``url_key`` that
+    users will request these contents from the cache. See `~download_file` for
+    details.
+
+    If ``url_key`` already exists in the cache, it will be updated to point to
+    these imported contents, and its old contents will be deleted if nothing
```

Comment:
Make it clear that deletion is optional, otherwise people might freak out.
```suggestion
    these imported contents, and if user so desires, its old contents will be deleted if nothing
```

Reply:
It's not optional. If you make `url_key` point to something new, you should expect the old thing to go away - there's no way to access it any more. The sole exception is if some other URL pointed to the same contents; then it won''t be deleted out from under that other URL. This behaviour is just like python dictionaries - when you change a value, the dictionary lets go of its reference to the old value. If nothing else points there, it's deallocated.

Reply:
Ah, gotcha. Sorry, I got confused between the cache content and the filename behaviors.